ENH Add extension for new signing key and upgrade to CMS 5

mikenuguid opened the pull request:
- Adds ability to set own signing key via `AuthorizationValidatorInterface` . This is also a refactor of the previous [approach](https://github.com/silverstripeltd/ss-oauth2-server/tree/feature/use-secure-eddsa) allowing any user to customise a signing key on a project. A default validator is used by default making this backwards compatible.
- Upgrade to work with CMS 5
- Revert modified unit tests
- Add new unit tests for extension points
- Add example docs for EdDSA

mikenuguid committed:
ENH Update to EDDSA(ED25519) signing key

mikenuguid committed:
ENH Refactor code and ReadMe

Jianbinzhu committed:
CMS 5 upgrade

Jianbinzhu committed:
Update monolog/monolog to v3

Jianbinzhu committed:
Update Sever controller

mikenuguid committed:
Add private properties to avoid deprecation notice

- Adding a value to dynamic properties is getting flagged by newer PHP versions e.g. 8.3, add this properties to the class itself since they are cannot be inherited from parent.

mikenuguid committed:
FIX unit tests

mikenuguid committed:
Refactor code

- use dependency injection to add new token validator for new algorithm
- add extension points for generating token
- remove unnecessary code from package

mikenuguid committed:
Add docs and revert unit tests

- add docs for configuring EdDSA/Ed25519 access token
- revert previously updated unit tests
- add unit tests for new API

mikenuguid commented:
[@blueo](https://github.com/blueo) would appreciate if I could have your feedback on this when you have time.

mikenuguid commented:
Hey @blueo, any chance you could take a look at this PR? 🙂

blueo commented on code/OauthServerController.php:
```diff
@@ -185,12 +192,22 @@ public function __construct()
+        return $this->authorizationValidator;
+    }
+
+    public function setAuthorizationValidator(?AuthorizationValidatorInterface $value): void
```
is there a use case for setting this to null? If not then you could probably remove `?` from the type

mikenuguid replied:
Yes. If null is passed, the `ResourceServer` will use `BearerTokenValidator` (tied up to RSA) by default.

blueo commented on tests/OauthServerControllerTest.php:
```diff
@@ -322,7 +324,16 @@ public function now(): DateTimeImmutable
-        OauthServerController::config()->merge('grant_expiry_interval', 'PT1H');
-        $this->assertSame('PT1H', $oauthController::getGrantTypeExpiryInterval());
+        OauthServerController::config()->merge('grant_expiry_interval', ['PT1H']);
+        $this->assertSame('PT1H', $oauthController::getGrantTypeExpiryInterval()[0]);
```
what's the rationale for loosening the type on `getGrantTypeExpiryInterval` and allowing an array? Seems a string is what you're using here anyway and its changing the API

mikenuguid replied:
Can't recall unfortunately, probably a failing test. I'll have a look. Thanks.

blueo commented on code/OauthServerController.php:
```diff
@@ -311,9 +329,9 @@ public static function authenticateRequest($controller): ?ServerRequestInterface
      * @param mixed $controller
      */
-    public static function getMember($controller): ?Member
+    public function getMember($controller): ?Member
```
this is a breaking change (which is fine so long as you're planning to publish a new version). But its also referenced on the main readme.md as how to use the module so it'd be good to update it there too

mikenuguid replied:
I'll double check this one too. But thanks. I'll make sure to update the doc, it's mentioned in there.

blueo commented on docs/eddsa.md:
````diff
@@ -0,0 +1,45 @@
+### Generate a private/public key pair using Ed25519 algorithm 
+
+```shell
+openssl genpkey -algorithm Ed25519 -out private.key
````
looks like we should probably put this as the recommended way on the main Readme.md? it is a better default than RSA

mikenuguid replied:
A recommentdation yes, but probably leave RSA as default since not most application use the new algorithm (more so support it)

blueo commented on code/Entities/AccessTokenEntity.php:
```diff
@@ -82,6 +86,34 @@ class AccessTokenEntity extends DataObject implements AccessTokenEntityInterface
+        $token = null;
+
+        // Get token from extension (in case of different implementation than the default)
+        $this->extend('updateJWT', $token);
```
i'm wondering about this change. Seems that most of the parts of `convertToJWT` are able to be changed in other public methods eg `getScopes` or you can override the public initJwtConfiguration and use the setters on https://lcobucci-jwt.readthedocs.io/en/stable/configuration/ ?  Do you have an example of an extension? I'm thinking eddsa is probably a better default it could be good to include that change but allow updating it with injector or similar

mikenuguid replied:
https://github.com/silverstripeltd/westpac-2020/blob/master/app/src/DAM/OauthServer/AccessTokenEntityExtension.php#L21

blueo replied:
yeah looking at that implementation it seems it would be better to override the public `initJwtConfiguration` rather than the private convertToJWT as that detail may change in future non breaking releases given its a private method. It also looks like that implementation could be part of the module? It could be configured using yaml instead of the default sha256 version as an option?   

https://github.com/thephpleague/oauth2-server/blob/8.5.x/src/Entities/Traits/AccessTokenTrait.php#L44

mikenuguid replied:
> It also looks like that implementation could be part of the module?

@blueo Double checking, Are you thinking about adding the eddsa config but still use RSA as default?

Given the amount of changes requested, I might need to ask for more time (budget) than we initially estimated this for so I might revisit it again some time in the future. Appreciate the feedback!

blueo replied:
yeah I'd suggest making eddsa the default and bringing that implementation from the internal module in here as rsa getting a bit old now AFAIK. but could also leave it as default with just docs on how to change. Yeah can leave the implementation to when you can get to it :)

blueo reviewed (changes requested):
I'm by no means an expert on this module but I've left a couple of questions on some implementation details

blueo reviewed (changes requested):
looking pretty good but I think the extension is probably on the wrong method and ideally we could include the Eddsa version of the configuration in the module too

mikenuguid committed:
FIX Update validation for grant type

- This should validate the grant type set on the CMS and the request. Previously, we set a different grant type on the CMS from the request.
- Note: Grant Type value on the CMS should be updated otherwise this change will invalidate the (access token) request if the grant types are not identical.

mikenuguid committed:
FIX Remove additional expiry time when issuing new access token

- The access token expiry is added by default when issuing a new token so adding up from config can be removed.

mikenuguid committed:
FIX unit tests

- revert some previous "fixes" since they are not necessary anymore after running phpunit on barebone silverstripe installation

mikenuguid commented:
Hi @blueo. It's been a while but we got budget to update this package. Just a heads up I'll be closing this PR in exchange for a CMS 6 update which includes the listed changes. 🙂

mikenuguid commented:
Closing. Please see latest pull request: https://github.com/IanSimpson/ss-oauth2-server/pull/32

@blueo ^
